tests/e2e/json_utils.go

```diff
@@ -0,0 +1,412 @@
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.downtimeSlashAction":
+		var a downtimeSlashAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.unbondTokensAction":
+		var a unbondTokensAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.cancelUnbondTokensAction":
+		var a cancelUnbondTokensAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.redelegateTokensAction":
+		var a redelegateTokensAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.doublesignSlashAction":
+		var a doublesignSlashAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.startRelayerAction":
+		var a startRelayerAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.slashThrottleDequeueAction":
+		var a slashThrottleDequeueAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.createIbcClientsAction":
+		var a createIbcClientsAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.LegacyUpgradeProposalAction":
+		var a LegacyUpgradeProposalAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.waitUntilBlockAction":
+		var a waitUntilBlockAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.ChangeoverChainAction":
+		var a ChangeoverChainAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.StartSovereignChainAction":
+		var a StartSovereignChainAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.lightClientEquivocationAttackAction":
+		var a lightClientEquivocationAttackAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.lightClientAmnesiaAttackAction":
+		var a lightClientAmnesiaAttackAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	case "main.lightClientLunaticAttackAction":
+		var a lightClientLunaticAttackAction
+		err := json.Unmarshal(rawAction, &a)
+		if err == nil {
+			return a, nil
+		}
+	default:
+		return nil, fmt.Errorf("unknown action type: %s", actionTypeString)
+	}
+	return nil, err
+}
+
+// custom marshal and unmarshal functions for the chainstate that convert proposals to/from the auxiliary type with type info
+
+// MarshalJSON transforms the ChainState into a ChainStateWithProposalTypes by adding type info to the proposals
+func (c ChainState) MarshalJSON() ([]byte, error) {
+	type ProposalAndType struct {
+		RawProposal interface{}
+		Type        string
+	}
+
+	type ChainStateWithProposalTypes struct {
```

Comment:
Same question

Reply:
Done. I pulled it out from the method. 05232d933